exporters/prometheus/exporter.go

```diff
@@ -0,0 +1,185 @@
+
+import (
+	"context"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+)
+
+// Exporter is a Prometheus exporter.
+type Exporter struct {
+	reader   metric.Reader
+	registry *prometheus.Registry
+}
+
+// NewPrometheusExporter returns a Prometheus exporter.
+func NewPrometheusExporter(reader metric.Reader) (*Exporter, error) {
+	e := &Exporter{
+		reader:   reader,
+		registry: prometheus.NewRegistry(),
+	}
+
+	if err := e.registry.Register(e); err != nil {
+		return nil, err
+	}
+
+	return e, nil
+}
+
+// Describe implements prometheus.Collector.
+func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
+	metrics, err := e.reader.Collect(context.TODO())
+	if err != nil {
+		otel.Handle(err)
+	}
+	for _, metricData := range getMetricData(metrics) {
+		ch <- metricData.description
+	}
+}
+
+// Collect implements prometheus.Collector.
+func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+	metrics, err := e.reader.Collect(context.TODO())
+	if err != nil {
+		otel.Handle(err)
+	}
+
+	for _, metricData := range getMetricData(metrics) {
+		if metricData.valueType == prometheus.UntypedValue {
+			m, err := prometheus.NewConstHistogram(metricData.description, metricData.histogramCount, metricData.histogramSum, metricData.histogramBuckets, metricData.attributeValues...)
+			if err != nil {
+				otel.Handle(err)
+			}
+			ch <- m
+		} else {
+			m, err := prometheus.NewConstMetric(metricData.description, metricData.valueType, metricData.value, metricData.attributeValues...)
+			if err != nil {
+				otel.Handle(err)
+			}
+			ch <- m
+		}
+	}
+}
+
+// metricData holds the metadata as well as values for individual data points.
+type metricData struct {
+	name             string
+	description      *prometheus.Desc
+	attributeValues  []string
+	valueType        prometheus.ValueType
+	value            float64
+	histogramCount   uint64
+	histogramSum     float64
+	histogramBuckets map[float64]uint64
+}
+
+func getMetricData(metrics metricdata.ResourceMetrics) []*metricData {
+	allMetrics := make([]*metricData, 0)
+	for _, scopeMetrics := range metrics.ScopeMetrics {
+		for _, m := range scopeMetrics.Metrics {
+			switch v := m.Data.(type) {
+			case metricdata.Histogram:
+				allMetrics = append(allMetrics, getHistogramMetricData(v, m)...)
+			case metricdata.Sum[int64]:
+				allMetrics = append(allMetrics, getSumMetricData(v, m)...)
+			case metricdata.Sum[float64]:
+				allMetrics = append(allMetrics, getSumMetricData(v, m)...)
+			case metricdata.Gauge[int64]:
+				allMetrics = append(allMetrics, getGaugeMetricData(v, m)...)
+			case metricdata.Gauge[float64]:
+				allMetrics = append(allMetrics, getGaugeMetricData(v, m)...)
+			}
+		}
+	}
+
+	return allMetrics
+}
+
+func getHistogramMetricData(histogram metricdata.Histogram, m metricdata.Metrics) []*metricData {
```

Comment:
we should drop histograms with delta aggregation temporality.

[Spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/data-model.md#histograms-1)

Reply:
Likely want to log this drop using the [global logger](https://pkg.go.dev/go.opentelemetry.io/otel@v1.9.0/internal/global#Info).